dbt/adapters/snowflake/impl.py

```diff
@@ -129,6 +139,83 @@ def get_columns_in_relation(self, relation):
             else:
                 raise
 
+    def _show_object_metadata(self, relation):
+        try:
+            kwargs = {"relation": relation}
+            results = self.execute_macro(SHOW_OBJECT_METADATA_MACRO_NAME, kwargs=kwargs)
+        except DbtDatabaseError as exc:
+            # If we don't have permissions to the object, return empty
+            if "Object does not exist" in str(exc):
+                return None
+        else:
+            # If we don't get any results, return empty
+            if len(results) == 0:
+                return None
+
+            return results
+
+    def get_catalog_for_single_relation(self, relation: BaseRelation) -> Optional[CatalogTable]:
+        object_metadata = self._show_object_metadata(relation)
+
+        if object_metadata:
+            row = object_metadata[0]
+
+            is_dynamic = row.get("is_dynamic")
+            kind = row.get("kind")
+
+            if is_dynamic == "YES" and kind == "BASE TABLE":
+                table_type = "DYNAMIC TABLE"
+            else:
+                table_type = kind
+
+            table_metadata = TableMetadata(
+                type=table_type,
+                schema=row.get("schema_name"),
+                name=row.get("name"),
+                database=row.get("database_name"),
+                comment=row.get("comment"),
+                owner=row.get("owner"),
+            )
+
+            stats_dict: StatsDict = {
+                "has_stats": StatsItem(
+                    id="has_stats",
+                    label="Has Stats?",
+                    value=True,
+                    include=False,
+                    description="Indicates whether there are statistics for this table",
+                ),
+                "row_count": StatsItem(
+                    id="row_count",
+                    label="Row Count",
+                    value=row.get("rows"),
+                    include=True,
+                    description="Number of rows in the table as reported by Snowflake",
+                ),
+                "bytes": StatsItem(
+                    id="bytes",
+                    label="Approximate Size",
+                    value=row.get("bytes"),
+                    include=True,
+                    description="Size of the table as reported by Snowflake",
+                ),
+            }
+
+            catalog_columns = {}
+            columns_metadata = self.get_columns_in_relation(relation)
+
+            for i, c in enumerate(columns_metadata):
```

Comment:
These lines could be written as a `dict` comprehension. I don't know if it's necessarily better, but for the sake of comparison:
```python
catalog_columns = {
    c.column: ColumnMetadata(type=c.dtype, index=i + 1, name=c.column)
    for i, c in enumerate(self.get_columns_in_relation(relation))
}
```